ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,314 @@
+    super.onResponse(replicaId, trackedRequestFinalState);
+    modifyReplicasInPoolOrInFlightCount(replicaId, -1);
+
+    String dcName = replicaId.getDataNodeId().getDatacenterName();
+
+    switch (trackedRequestFinalState) {
+      case SUCCESS:
+        replicaSuccessCountByDc.put(dcName, replicaSuccessCountByDc.get(dcName) + 1);
+        if (dcName.equals(datacenterName)) {
+          localReplicaSuccessCount++;
+        } else {
+          remoteReplicaSuccessCount++;
+        }
+        break;
+      case REQUEST_DISABLED:
+        disabledCountByDc.put(dcName, disabledCountByDc.get(dcName) + 1);
+        break;
+      default:
+        failedCountByDc.put(dcName, failedCountByDc.get(dcName) + 1);
+    }
+  }
+
+  public boolean hasFailed() {
+    if (routerConfig.routerPutUseDynamicSuccessTarget) {
+      for(String dcName : allDatacenters) {
+        if (dcName.equals(datacenterName)) {  // Local colo
+          if (hasFailedDynamically(dcName, routerConfig.routerPutSuccessTarget)) return true;
+        } else {
+          if (hasFailedDynamically(dcName, routerConfig.routerPutRemoteSuccessTarget)) return true;
+        }
+      }
+      return false;
+    } else {
+      return hasFailedLocally() || hasFailedRemotely();
+    }
+  }
+
+  private boolean hasFailedDynamically(String dc, int successTarget)  {
+    return (totalReplicaCountByDc.get(dc) - failedCountByDc.get(dc)) < Math.max(totalReplicaCountByDc.get(dc) - 1, successTarget + disabledCountByDc.get(dc));
+  }
+
+  private boolean hasFailedRemotely() {
+    for(String dcName : allDatacenters) {
+      if(!dcName.equals(datacenterName)) {
+        if (replicaSuccessCountByDc.get(dcName) < Math.max(totalReplicaCountByDc.get(dcName) - 1,
+            routerConfig.routerPutRemoteSuccessTarget + disabledCountByDc.get(dcName))) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private boolean hasFailedLocally() {
+    return replicaInPoolOrFlightCountByDc.get(datacenterName) + replicaSuccessCountByDc.get(datacenterName) < replicaSuccessTarget;
+  }
+
+  /**
+   * Add {@code delta} to a replicas in pool or in flight counter.
+   * @param delta the value to add to the counter.
+   */
+  private void modifyReplicasInPoolOrInFlightCount(ReplicaId replica, int delta) {
+    String dcName = replica.getDataNodeId().getDatacenterName();
+    replicaInPoolOrFlightCountByDc.put(dcName, replicaInPoolOrFlightCountByDc.get(dcName) + delta);
+  }
+
+
+  int getCurrentLocalParallelism() {
+    return replicaParallelism;
+  }
+
+  int getCurrentRemoteParallelism() {
+    return remoteReplicaParallelism;
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    return otIterator;
+  }
+
+  private class ParanoidDurabilityTrackerIterator implements Iterator<ReplicaId> {
+    private int currentDatacenterIndex = 0;
+
+    @Override
+    public boolean hasNext() {
+      return hasNextLocal() || hasNextRemote();
```

Comment:
I think we should formalize the order of each replica somewhere so there is no confusion.

The idea is that we should try local datacenter first, make sure local datacenter succeeds, then we move on to remote. If this is true, then the logic of `hasNext` should look something like
```
return hasNextLocal() || (localFinished() && hasNextRemote());
```

This is important. For example, if the local parallelism is 2 and we have 3 local replicas, then 
```
Iterator<ReplicaId> iter = operationTracker.getReplicaIterator();
iter.hasNext(); // should return true;
ReplicaId r1 = iter.next();
iter.remove();
iter.hasNext(); // should return true;
ReplicaId r2 = iter.next();
iter.remove();
it.hasNext() // should return false, we only need two parallel local replicas.
```
We should not return true on third `hasNext` call, since the local replicas are not yet finished.

Reply:
We talked about this at length in a meeting and agreed to keep this part of the PR as it exists. The idea is that we fire off requests to the local colo as well as the remote colos in parallel (while staying within the configured bounds), in order to reduce the full latency as seen by our clients.